File: BOJ/29_유니온파인드/1976_여행가자.cpp
#include <iostream>

#define MAXN 200
#define MAXM 1000
using namespace std;
int adjs[MAXN+1][MAXN+1], par[MAXN+1], path[MAXM+1];
int N,M,x;

int find(int x) {
    if(par[x] == x) return par[x];
    else return par[x] = find(par[x]);
}

void merge(int x, int y) {
    int p_x = find(x);
    int p_y = find(y);
    if(p_x == p_y) return ;
    par[p_x] = p_y;
}

int main() {
    scanf("%d %d",&N,&M);
    for(int i=0; i<=N; i++) par[i] = i;
    for(int i=1; i<=N; i++) {
        for(int j=1; j<=N; j++) {
            scanf("%d",&x);
            if(x == 1) merge(i,j);
        }
    }
    for(int i=0; i<M; i++) {
        scanf("%d",path+i);
    }
    for(int i=1; i<M; i++) {
        if(find(path[i-1]) != find(path[i])) {
            printf("NO\n");
            return 0;
        }
    }
    printf("YES\n");
    return 0;

}

File: BOJ/29_유니온파인드/1976_여행가자.py
import sys

read = sys.stdin.readline

def find(x) :
    if par[x] == x : return par[x]
    else :
        par[x] = find(par[x])
        return par[x]

def merge(x,y) :
    p_x = find(x)
    p_y = find(y)

    if p_x == p_y : return
    par[p_x] = p_y

N = int(read())
M = int(read())
par = [i for i in range(N+1)]
adjs = []
for _ in range(N) :
    adjs.append(list(map(int,read().split())))
path = list(map(int,read().split()))

for i in range(N) :
    for j in range(N) :
        if adjs[i][j] == 1 :
            merge(i+1,j+1)

for i in range(1,len(path)) :
    if find(path[i-1]) != find(path[i]) :
        print("NO")
        sys.exit()

print("YES")